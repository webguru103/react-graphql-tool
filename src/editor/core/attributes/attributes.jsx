import * as React from 'react';
import { InputBase, withStyles } from '@material-ui/core';
import IconButton from '@material-ui/core/IconButton';
import DeleteIcon from '../../../assets/trash.svg';
import CopyIcon from '../../../assets/copy.svg';
import styles from './attributes.styles';

type PropType = {
  classes: Object,
  fieldName: string,
  fieldType: string,
  children: React.Node,
  activePageFieldId: number,
  pageIndex: number,
  handleFieldUpdate: Function,
  handleFieldClone: Function,
  handleDeleteFields: Function,
  disabled: boolean,
};

type StateType = {
  activePageFieldId: number,
  fieldName: string,
};

class AttributesC extends React.PureComponent<PropType, StateType> {

  state = {
    activePageFieldId: this.props.activePageFieldId,
    fieldName: this.props.fieldName || '',
  };

  componentWillUnmount() {
    const { fieldType, pageIndex, handleFieldUpdate } = this.props;
    const { activePageFieldId } = this.state;

    // This is to ensure that any updates to field name is saved when clicking outside of the panel and field blur is not triggered
    // If the props field name is different from the state field name, new name has not been saved.
    if (this.props.fieldName !== this.state.fieldName) {
      handleFieldUpdate([
        {
          id: activePageFieldId,
          fieldName: this.state.fieldName,
          type: fieldType,
          pageIndex,
        },
      ]);
    }
  }

  updateFieldName = (ev: SyntheticEvent<HTMLInputElement>) => {
    const { value } = ev.currentTarget;

    this.setState({
      fieldName: value,
    });
  };

  render() {
    const {
      classes,
      children,
      activePageFieldId,
      fieldType,
      pageIndex,
      handleFieldUpdate,
      handleFieldClone,
      handleDeleteFields,
      disabled,
    } = this.props;

    return (
      <div className={classes.root}>
        <div className={classes.header}>
          <div className={classes.headerTitle}>
            <InputBase
              id="fieldName"
              fullWidth
              onChange={ev => this.updateFieldName(ev)}
              className={classes.headerTitleInput}
              onBlur={() => {
                if (this.props.fieldName !== this.state.fieldName) {
                  handleFieldUpdate([
                    {
                      id: activePageFieldId,
                      fieldName: this.state.fieldName,
                      type: fieldType,
                      pageIndex,
                    },
                  ]);
                }
              }}
              value={this.state.fieldName}
              disabled={disabled}
            />
          </div>
          <div className={classes.headerIcons}>
            <IconButton className={classes.editIcon} onClick={() => handleFieldClone()} disabled={disabled} data-testid="cloneField">
              <CopyIcon />
            </IconButton>
            <IconButton className={classes.editIcon} onClick={() => handleDeleteFields()} disabled={disabled} data-testid="deleteField">
              <DeleteIcon />
            </IconButton>
          </div>
        </div>
        <div className={classes.content}>
          {children}
        </div>
      </div>
    );
  }

}

export default withStyles(styles, { withTheme: true })(AttributesC);
