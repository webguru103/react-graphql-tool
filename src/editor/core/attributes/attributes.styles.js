export const ATTRIBUTE_PANEL_MIN_WIDTH = 250;

export default (theme: Object) => ({
  root: {
    padding: '17px 20px 0 25px',
    height: '100%',
    overflow: 'hidden',
    minWidth: ATTRIBUTE_PANEL_MIN_WIDTH,
  },
  header: {
    display: 'flex',
    minHeight: 40,
    alignItems: 'center',
    background: theme.palette.neutrals.background,
  },
  headerTitle: {
    flexGrow: '2',
  },
  headerTitleInput: {
    fontSize: 19,
    fontWeight: 600,
  },
  headerIcons: {
    display: 'flex',
    flexFlow: 'row',
  },
  content: {
    height: '100%',
    background: theme.palette.neutrals.background,
  },
});
