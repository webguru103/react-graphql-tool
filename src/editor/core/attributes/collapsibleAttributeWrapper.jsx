import * as React from 'react';
import ExpansionPanel from '@material-ui/core/ExpansionPanel';
import { withStyles } from '@material-ui/core';
import ExpansionPanelSummary from '@material-ui/core/ExpansionPanelSummary';
import ExpansionPanelDetails from '@material-ui/core/ExpansionPanelDetails';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';
import Typography from '@material-ui/core/Typography';
import classNames from 'classnames';
import styles from './collapsibleAttributeWrapper.styles';

type PropType = {
  attributeName: string,
  children: React.Element,
  classes: Object,
};

type StateType = {
  expanded: boolean,
}

class CollapsibleAttributeWrapperC extends React.Component<PropType, StateType> {

  state = {
    expanded: true,
  };

  handleExpand = () => {
    this.setState(prevState => ({ expanded: !prevState.expanded }));
  };

  render() {
    const { expanded } = this.state;

    const { attributeName, children, classes } = this.props;
    return (
      <ExpansionPanel
        expanded={expanded}
        CollapseProps={{ classes: { container: classNames(classes.panel, { [classes.panelExpanded]: expanded }) } }}
        onChange={this.handleExpand}
        elevation={0}
      >
        <ExpansionPanelSummary
          classes={{
            root: classes.root, expanded: classes.expanded, content: classes.content, expandIcon: classes.expandIcon,
          }}
          expandIcon={<ExpandMoreIcon />}
        >
          <Typography className={classes.attributeName}>{attributeName}</Typography>
        </ExpansionPanelSummary>
        <ExpansionPanelDetails classes={{ root: classes.expansionPanelDetails }}>
          {children}
        </ExpansionPanelDetails>
      </ExpansionPanel>
    );
  }

}

export default withStyles(styles)(CollapsibleAttributeWrapperC);
