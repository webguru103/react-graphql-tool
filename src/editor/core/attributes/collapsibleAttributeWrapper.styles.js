export default () => ({
  root: {
    marginBottom: -1,
    minHeight: 50,
    '&$expanded': {
      minHeight: 50,
    },
    padding: 0,
  },
  content: {
    '&$expanded': {
      margin: '12px 0px',
    },
  },
  expanded: {},
  expandIcon: {
    padding: 0,
  },
  panel: {
    overflow: 'hidden',
  },
  panelExpanded: {
    overflow: 'visible',
  },
  expansionPanelDetails: {
    padding: 0,
  },
  attributeName: {
    fontSize: 15,
    fontWeight: 600,
  },
});
