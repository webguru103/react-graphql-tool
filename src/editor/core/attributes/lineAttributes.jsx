import * as React from 'react';
import debounce from 'lodash.debounce';

import FormLabel from '@material-ui/core/FormLabel';
import Input from '@material-ui/core/Input';
import { withStyles } from '@material-ui/core';
import InputLabel from '@material-ui/core/InputLabel';
import Checkbox from '@material-ui/core/Checkbox';
import Select from 'react-select';

import { LINE_MIN_STROKE_THICKNESS, LINE_MAX_STROKE_THICKNESS } from '../../constants';
import styles from './lineAttributes.styles';
import Attributes from './attributes';
import CollapsibleAttributeWrapper from './collapsibleAttributeWrapper';
import type { AdaptedLineField } from '../flowTypes';
import { adjustLineAttributesForPageSize } from '../utils';

type Props = {
  classes: Object,
  field: AdaptedLineField,
  handleLineFieldUpdate: Function,
  pageRef: HTMLDivElement,
  disabled: boolean,
  zoom: number,
  pageIndex: number,
};

type State = {
  x1: number,
  y1: number,
  x2: number,
  y2: number,
  positionLock: boolean,
  strokeThickness: number,
};

export class LineAttributesC extends React.Component<Props, State> {

  state = {
    x1: this.props.field.x1 || 0,
    y1: this.props.field.y1 || 0,
    x2: this.props.field.x2 || 0,
    y2: this.props.field.y2 || 0,
    positionLock: this.props.field.positionLock || false,
    strokeThickness: this.props.field.strokeThickness || 1,
  };

  updateLineField = debounce((currentState: State) => {
    const { handleLineFieldUpdate, field, pageIndex } = this.props;
    const {
      x1,
      y1,
      x2,
      y2,
      positionLock,
      strokeThickness,
    } = currentState;

    const lineField = { ...field };

    if (!field.positionLock) {
      lineField.x1 = x1;
      lineField.y1 = y1;
      lineField.x2 = x2;
      lineField.y2 = y2;
    }
    lineField.positionLock = positionLock;
    lineField.strokeThickness = strokeThickness;
    handleLineFieldUpdate(
      field.id,
      lineField,
      pageIndex,
    );
  }, 50)

  componentDidUpdate(prevProps: Props) {
    const {
      field: {
        x1, y1, x2, y2, strokeThickness,
      },
    } = this.props;
    if (prevProps.field.x1 !== x1
      || prevProps.field.y1 !== y1
      || prevProps.field.x2 !== x2
      || prevProps.field.y2 !== y2
      || prevProps.field.strokeThickness !== strokeThickness
    ) {
      this.setState({
        x1: x1 || 0,
        y1: y1 || 0,
        x2: x2 || 0,
        y2: y2 || 0,
        strokeThickness: strokeThickness || prevProps.field.strokeThickness,
      });
    }
  }

  handleChangePositionSize = (field: string, forceParentUpdate?: boolean) => (e: SyntheticInputEvent<HTMLInputElement>) => {
    const { pageRef, zoom } = this.props;
    const { positionLock, strokeThickness } = this.state;

    if (positionLock) {
      return;
    }

    const numberVal = e.target.value ? parseInt(e.target.value, 10) : 0;

    const update = {
      [field]: numberVal,
    };

    this.setState(update);

    const pageRect = pageRef.getBoundingClientRect();
    const adjusted = adjustLineAttributesForPageSize(pageRect.height / zoom, pageRect.width / zoom, update);

    if ((adjusted[field] === numberVal) || forceParentUpdate) {
      const adjustedUpdate = {
        [field]: adjusted[field],
      };
      this.setState(adjustedUpdate, () => {
        const currentState = {
          x1: this.state.x1,
          y1: this.state.y1,
          x2: this.state.x2,
          y2: this.state.y2,
          positionLock,
          strokeThickness,
        };
        this.updateLineField(currentState);
      });
    }
  };

  handleChangeStrokeThickness = (selection: {value: number, label: number}) => {
    const {
      x1,
      y1,
      x2,
      y2,
      positionLock,
    } = this.state;
    const numberVal = selection.value ? parseInt(selection.value, 10) : LINE_MIN_STROKE_THICKNESS;

    const currentState = {
      x1,
      y1,
      x2,
      y2,
      positionLock,
      strokeThickness: numberVal,
    };
    this.setState({ strokeThickness: numberVal }, () => this.updateLineField(currentState));
  };

  handlePositionLockToggle = () => {
    const update = {
      positionLock: !this.state.positionLock,
    };
    const {
      x1,
      y1,
      x2,
      y2,
      strokeThickness,
    } = this.state;

    const currentState = {
      x1,
      y1,
      x2,
      y2,
      positionLock: !this.state.positionLock,
      strokeThickness,
    };
    this.setState(update, () => this.updateLineField(currentState));
  };

  render() {
    const {
      classes, disabled, ...restProps
    } = this.props;
    const {
      x1,
      y1,
      x2,
      y2,
      positionLock,
      strokeThickness,
    } = this.state;

    return (
      <Attributes
        {...restProps}
        disabled={disabled}
        fieldName={this.props.field.fieldName}
        fieldType={this.props.field.type}
      >
        <React.Fragment>
          <CollapsibleAttributeWrapper attributeName="Position & Sizing">
            <div className={classes.section}>
              <div className={classes.sectionRow}>
                <FormLabel className={classes.formLabel}>Lock</FormLabel>
                <Checkbox
                  checked={positionLock}
                  onChange={this.handlePositionLockToggle}
                  value="updateFormBase"
                  name="updateFormBase"
                  disabled={disabled}
                />
              </div>
              <div className={classes.sectionRow}>
                <FormLabel className={classes.formLabel}>Vertex Point 1</FormLabel>
                <InputLabel classes={{ root: classes.inputLabel }}>x1</InputLabel>
                <Input
                  disabled={positionLock || disabled}
                  onChange={this.handleChangePositionSize('x1')}
                  onBlur={this.handleChangePositionSize('x1', true)}
                  classes={{ root: classes.input }}
                  value={x1.toString()}
                  type="number"
                  inputProps={{
                    'data-testid': 'line-attribute-x1-input',
                    min: 0,
                  }}
                />
                <InputLabel classes={{ root: classes.inputLabel }}>y1</InputLabel>
                <Input
                  disabled={positionLock || disabled}
                  onChange={this.handleChangePositionSize('y1')}
                  onBlur={this.handleChangePositionSize('y1', true)}
                  classes={{ root: classes.input }}
                  value={y1.toString()}
                  inputProps={{ min: 0 }}
                  type="number"
                />
              </div>
              <div className={classes.sectionRow}>
                <FormLabel className={classes.formLabel}>Vertex Point 2</FormLabel>
                <InputLabel classes={{ root: classes.inputLabel }}>x2</InputLabel>
                <Input
                  disabled={positionLock || disabled}
                  onChange={this.handleChangePositionSize('x2')}
                  onBlur={this.handleChangePositionSize('x2', true)}
                  classes={{ root: classes.input }}
                  value={x2.toString()}
                  inputProps={{ min: 0 }}
                  type="number"
                />
                <InputLabel classes={{ root: classes.inputLabel }}>y2</InputLabel>
                <Input
                  disabled={positionLock || disabled}
                  onChange={this.handleChangePositionSize('y2')}
                  onBlur={this.handleChangePositionSize('y2', true)}
                  classes={{ root: classes.input }}
                  value={y2.toString()}
                  inputProps={{ min: 0 }}
                  type="number"
                />
              </div>
            </div>
          </CollapsibleAttributeWrapper>
          <CollapsibleAttributeWrapper attributeName="Stroke Appearance">
            <div className={classes.section} data-testid="line-attribute-x2-input">
              <div className={classes.sectionRow}>
                <FormLabel classes={{ root: classes.formLabel }}>Thickness</FormLabel>
                <Select
                  className={classes.select}
                  value={{ value: strokeThickness, label: strokeThickness }}
                  autosize
                  onChange={this.handleChangeStrokeThickness}
                  options={[...Array(LINE_MAX_STROKE_THICKNESS).keys()].map(k => k + 1).map(thickness => ({ value: thickness, label: thickness }))}
                  isDisabled={disabled}
                />
              </div>
            </div>
          </CollapsibleAttributeWrapper>
        </React.Fragment>
      </Attributes>
    );
  }

}

export default withStyles(styles, { withTheme: true })(LineAttributesC);
