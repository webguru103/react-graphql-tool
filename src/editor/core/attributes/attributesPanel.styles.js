import { ATTRIBUTE_PANEL_MIN_WIDTH } from './attributes.styles';

export default () => ({
  root: {
    overflow: 'scroll',
  },
  noSelection: {
    minWidth: ATTRIBUTE_PANEL_MIN_WIDTH,
    display: 'flex',
    height: '100%',
    alignItems: 'center',
    justifyContent: 'center',
  },
});
