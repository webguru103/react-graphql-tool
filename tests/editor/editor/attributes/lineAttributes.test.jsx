/* eslint import/no-extraneous-dependencies: 0 */

import * as React from 'react';
import 'jest-dom/extend-expect';
import { render, cleanup, fireEvent } from 'react-testing-library';
import { withContext } from '../../../test-helpers/contextWrapper';
import { LineAttributesC } from '../../../../src/editor/core/attributes/lineAttributes';

describe('LineField', () => {

  jest.useFakeTimers();

  afterEach(cleanup);
  const Component = withContext(LineAttributesC);

  test('renders', () => {

    const field = {
      id: 53151,
      fieldName: 'LineField',
      x1: 10,
      y1: 10,
      x2: 10,
      y2: 30,
      strokeThickness: 5,
    };

    const pageRef = {
      getBoundingFieldRect: () => {},
    };

    const handleLineFieldUpdate = jest.fn();
    const activePageFieldPageNumber = 0;

    const { container } = render(<Component
      classes={{}}
      field={field}
      handleLineFieldUpdate={handleLineFieldUpdate}
      activePageFieldPageNumber={activePageFieldPageNumber}
      pageRef={pageRef}
    />);

    expect(container).toMatchSnapshot();
  });

  test('calls handleLineFieldUpdate with correct input', () => {

    const field = {
      id: 53151,
      fieldName: 'LineField',
      x1: 10,
      y1: 10,
      x2: 10,
      y2: 30,
      strokeThickness: 5,
      pageNumber: 0,
      positionLock: false,
    };

    const pageRef = {
      getBoundingClientRect: () => ({
        height: 500,
        width: 600,
      }),
    };

    const newX1Input = '500';

    const handleLineFieldUpdate = jest.fn();
    const activePage = 0;

    const { getByTestId } = render(<Component
      classes={{}}
      field={field}
      handleLineFieldUpdate={handleLineFieldUpdate}
      pageIndex={activePage}
      pageRef={pageRef}
    />);

    const x1Input = getByTestId('line-attribute-x1-input');
    fireEvent.change(x1Input, { target: { value: newX1Input } });

    const x2Input = getByTestId('line-attribute-x2-input');
    fireEvent.click(x2Input);

    // Fast forward all timers so Loading indicator content is rendered
    jest.runAllTimers();

    const expected = {
      fieldName: 'LineField',
      id: 53151,
      positionLock: false,
      strokeThickness: 5,
      x1: Number(newX1Input),
      x2: 10,
      y1: 10,
      y2: 30,
      pageNumber: 0,
    };

    expect(handleLineFieldUpdate.mock.calls.length).toBe(1);
    expect(handleLineFieldUpdate).lastCalledWith(
      field.id,
      expected,
      activePage,
    );
  });

  test('calls handleLineFieldUpdate with correct input after adjusting to page size', () => {

    const field = {
      id: 53151,
      fieldName: 'LineField',
      x1: 10,
      y1: 10,
      x2: 10,
      y2: 30,
      strokeThickness: 5,
      pageNumber: 0,
      positionLock: false,
    };

    const pageRef = {
      getBoundingClientRect: () => ({
        height: 500,
        width: 600,
      }),
    };

    const newX1Input = '800';
    const expectedWidthMax = '600';

    const handleLineFieldUpdate = jest.fn();
    const activePage = 0;

    const { getByTestId } = render(<Component
      classes={{}}
      field={field}
      handleLineFieldUpdate={handleLineFieldUpdate}
      pageIndex={activePage}
      pageRef={pageRef}
      zoom={1}
    />);

    const x1Input = getByTestId('line-attribute-x1-input');
    fireEvent.change(x1Input, { target: { value: newX1Input } });
    fireEvent.blur(x1Input);

    // Fast forward all timers so Loading indicator content is rendered
    jest.runAllTimers();

    const expected = {
      fieldName: 'LineField',
      id: 53151,
      positionLock: false,
      strokeThickness: 5,
      x1: Number(expectedWidthMax),
      x2: field.x2,
      y1: field.y1,
      y2: field.y2,
      pageNumber: 0,
    };

    expect(handleLineFieldUpdate.mock.calls.length).toBe(1);
    expect(handleLineFieldUpdate).lastCalledWith(
      field.id,
      expected,
      activePage,
    );
  });

});
