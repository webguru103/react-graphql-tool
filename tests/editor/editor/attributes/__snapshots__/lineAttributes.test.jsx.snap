// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LineField renders 1`] = `
<div>
  <div
    class="AttributesC-root-1"
  >
    <div
      class="AttributesC-header-2"
    >
      <div
        class="AttributesC-headerTitle-3"
      >
        <div
          class="MuiInputBase-root-7 MuiInputBase-fullWidth-16 AttributesC-headerTitleInput-4"
        >
          <input
            class="MuiInputBase-input-17"
            id="fieldName"
            type="text"
            value="LineField"
          />
        </div>
      </div>
      <div
        class="AttributesC-headerIcons-5"
      >
        <button
          class="MuiButtonBase-root-30 MuiIconButton-root-24"
          data-testid="cloneField"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label-29"
          >
            <svg
              height="20"
              viewBox="0 0 20 20"
              width="20"
              xmlns="http://www.w3.org/2000/svg"
              xmlns:xlink="http://www.w3.org/1999/xlink"
            >
              <defs>
                <path
                  d="M7.786 2.119c-.894 0-1.62.725-1.62 1.62v6.475c0 .894.726 1.62 1.62 1.62h6.476c.894 0 1.619-.726 1.619-1.62V3.738c0-.894-.725-1.619-1.62-1.619H7.787zm5.666 11.333v.81a3.238 3.238 0 0 1-3.238 3.238H3.738A3.238 3.238 0 0 1 .5 14.262V7.786a3.238 3.238 0 0 1 3.238-3.238h.81v-.81A3.238 3.238 0 0 1 7.786.5h6.476A3.238 3.238 0 0 1 17.5 3.738v6.476a3.238 3.238 0 0 1-3.238 3.238h-.81zm-1.619 0H7.786a3.238 3.238 0 0 1-3.238-3.238V6.167h-.81c-.894 0-1.619.725-1.619 1.619v6.476c0 .894.725 1.619 1.62 1.619h6.475c.894 0 1.62-.725 1.62-1.62v-.809z"
                  id="a"
                />
              </defs>
              <g
                fill="none"
                fill-rule="evenodd"
                transform="translate(1 1)"
              >
                <mask
                  fill="#fff"
                  id="b"
                >
                  <use
                    xlink:href="#a"
                  />
                </mask>
                <use
                  fill-rule="nonzero"
                  xlink:href="#a"
                />
                <g
                  fill="currentColor"
                  mask="url(#b)"
                >
                  <path
                    d="M-1-1h20v20H-1z"
                  />
                </g>
              </g>
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root-33"
          />
        </button>
        <button
          class="MuiButtonBase-root-30 MuiIconButton-root-24"
          data-testid="deleteField"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label-29"
          >
            <svg
              height="20"
              viewBox="0 0 20 20"
              width="20"
              xmlns="http://www.w3.org/2000/svg"
              xmlns:xlink="http://www.w3.org/1999/xlink"
            >
              <defs>
                <path
                  d="M4.444 3.367V1.589c0-.491.398-.889.89-.889h5.333c.49 0 .889.398.889.889v1.778h3.555a.889.889 0 0 1 0 1.777v10.667a3.556 3.556 0 0 1-3.555 3.556H4.444A3.556 3.556 0 0 1 .89 15.81V5.144a.889.889 0 1 1 0-1.777h3.555zM2.667 5.144v10.667c0 .982.796 1.778 1.777 1.778h7.112c.981 0 1.777-.796 1.777-1.778V5.144H2.667zm2.666 2.667a.889.889 0 1 1 1.778 0v7.111a.889.889 0 0 1-1.778 0v-7.11zm3.556 0a.889.889 0 0 1 1.778 0v7.111a.889.889 0 0 1-1.778 0v-7.11zM6.222 3.367h3.556v-.89H6.222v.89z"
                  id="a"
                />
              </defs>
              <g
                fill="none"
                fill-rule="evenodd"
                transform="translate(2)"
              >
                <mask
                  fill="#fff"
                  id="b"
                >
                  <use
                    xlink:href="#a"
                  />
                </mask>
                <use
                  fill-rule="nonzero"
                  xlink:href="#a"
                />
                <g
                  fill="currentColor"
                  mask="url(#b)"
                >
                  <path
                    d="M-2 0h20v20H-2z"
                  />
                </g>
              </g>
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root-33"
          />
        </button>
      </div>
    </div>
    <div
      class="AttributesC-content-6"
    >
      <div
        class="MuiPaper-root-51 MuiPaper-elevation0-53 MuiExpansionPanel-root-48 MuiExpansionPanel-expanded-49"
      >
        <div
          aria-expanded="true"
          class="MuiButtonBase-root-30 MuiExpansionPanelSummary-root-78 CollapsibleAttributeWrapperC-root-40 MuiExpansionPanelSummary-expanded-79 CollapsibleAttributeWrapperC-expanded-42"
          role="button"
          tabindex="0"
        >
          <div
            class="MuiExpansionPanelSummary-content-82 CollapsibleAttributeWrapperC-content-41 MuiExpansionPanelSummary-expanded-79 CollapsibleAttributeWrapperC-expanded-42"
          >
            <p
              class="MuiTypography-root-84 MuiTypography-body1-93 CollapsibleAttributeWrapperC-attributeName-47"
            >
              Position & Sizing
            </p>
          </div>
          <div
            aria-hidden="true"
            class="MuiButtonBase-root-30 MuiIconButton-root-24 MuiExpansionPanelSummary-expandIcon-83 CollapsibleAttributeWrapperC-expandIcon-43 MuiExpansionPanelSummary-expanded-79 CollapsibleAttributeWrapperC-expanded-42"
            role="button"
            tabindex="-1"
          >
            <span
              class="MuiIconButton-label-29"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root-110"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                />
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root-33"
            />
          </div>
        </div>
        <div
          class="MuiCollapse-container-119 CollapsibleAttributeWrapperC-panel-44 CollapsibleAttributeWrapperC-panelExpanded-45 MuiCollapse-entered-120"
          style="min-height: 0px;"
        >
          <div
            class="MuiCollapse-wrapper-121"
          >
            <div
              class="MuiCollapse-wrapperInner-122"
            >
              <div
                class="MuiExpansionPanelDetails-root-123 CollapsibleAttributeWrapperC-expansionPanelDetails-46"
              >
                <div>
                  <div>
                    <label
                      class="MuiFormLabel-root-124"
                    >
                      Lock
                    </label>
                    <span
                      class="MuiButtonBase-root-30 MuiIconButton-root-24 MuiSwitchBase-root-137 MuiCheckbox-root-131 MuiCheckbox-colorSecondary-136"
                    >
                      <span
                        class="MuiIconButton-label-29"
                      >
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root-110"
                          focusable="false"
                          role="presentation"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                          />
                        </svg>
                        <input
                          class="MuiSwitchBase-input-140"
                          data-indeterminate="false"
                          name="updateFormBase"
                          type="checkbox"
                          value="updateFormBase"
                        />
                      </span>
                      <span
                        class="MuiTouchRipple-root-33"
                      />
                    </span>
                  </div>
                  <div>
                    <label
                      class="MuiFormLabel-root-124"
                    >
                      Vertex Point 1
                    </label>
                    <label
                      class="MuiFormLabel-root-124 MuiInputLabel-root-141 MuiInputLabel-animated-145"
                    >
                      x1
                    </label>
                    <div
                      class="MuiInputBase-root-7 MuiInput-root-148 MuiInput-underline-152"
                    >
                      <input
                        class="MuiInputBase-input-17 MuiInput-input-156 MuiInputBase-inputType-20 MuiInput-inputType-159"
                        data-testid="line-attribute-x1-input"
                        min="0"
                        type="number"
                        value="10"
                      />
                    </div>
                    <label
                      class="MuiFormLabel-root-124 MuiInputLabel-root-141 MuiInputLabel-animated-145"
                    >
                      y1
                    </label>
                    <div
                      class="MuiInputBase-root-7 MuiInput-root-148 MuiInput-underline-152"
                    >
                      <input
                        class="MuiInputBase-input-17 MuiInput-input-156 MuiInputBase-inputType-20 MuiInput-inputType-159"
                        min="0"
                        type="number"
                        value="10"
                      />
                    </div>
                  </div>
                  <div>
                    <label
                      class="MuiFormLabel-root-124"
                    >
                      Vertex Point 2
                    </label>
                    <label
                      class="MuiFormLabel-root-124 MuiInputLabel-root-141 MuiInputLabel-animated-145"
                    >
                      x2
                    </label>
                    <div
                      class="MuiInputBase-root-7 MuiInput-root-148 MuiInput-underline-152"
                    >
                      <input
                        class="MuiInputBase-input-17 MuiInput-input-156 MuiInputBase-inputType-20 MuiInput-inputType-159"
                        min="0"
                        type="number"
                        value="10"
                      />
                    </div>
                    <label
                      class="MuiFormLabel-root-124 MuiInputLabel-root-141 MuiInputLabel-animated-145"
                    >
                      y2
                    </label>
                    <div
                      class="MuiInputBase-root-7 MuiInput-root-148 MuiInput-underline-152"
                    >
                      <input
                        class="MuiInputBase-input-17 MuiInput-input-156 MuiInputBase-inputType-20 MuiInput-inputType-159"
                        min="0"
                        type="number"
                        value="30"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiPaper-root-51 MuiPaper-elevation0-53 MuiExpansionPanel-root-48 MuiExpansionPanel-expanded-49"
      >
        <div
          aria-expanded="true"
          class="MuiButtonBase-root-30 MuiExpansionPanelSummary-root-78 CollapsibleAttributeWrapperC-root-40 MuiExpansionPanelSummary-expanded-79 CollapsibleAttributeWrapperC-expanded-42"
          role="button"
          tabindex="0"
        >
          <div
            class="MuiExpansionPanelSummary-content-82 CollapsibleAttributeWrapperC-content-41 MuiExpansionPanelSummary-expanded-79 CollapsibleAttributeWrapperC-expanded-42"
          >
            <p
              class="MuiTypography-root-84 MuiTypography-body1-93 CollapsibleAttributeWrapperC-attributeName-47"
            >
              Stroke Appearance
            </p>
          </div>
          <div
            aria-hidden="true"
            class="MuiButtonBase-root-30 MuiIconButton-root-24 MuiExpansionPanelSummary-expandIcon-83 CollapsibleAttributeWrapperC-expandIcon-43 MuiExpansionPanelSummary-expanded-79 CollapsibleAttributeWrapperC-expanded-42"
            role="button"
            tabindex="-1"
          >
            <span
              class="MuiIconButton-label-29"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root-110"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                />
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root-33"
            />
          </div>
        </div>
        <div
          class="MuiCollapse-container-119 CollapsibleAttributeWrapperC-panel-44 CollapsibleAttributeWrapperC-panelExpanded-45 MuiCollapse-entered-120"
          style="min-height: 0px;"
        >
          <div
            class="MuiCollapse-wrapper-121"
          >
            <div
              class="MuiCollapse-wrapperInner-122"
            >
              <div
                class="MuiExpansionPanelDetails-root-123 CollapsibleAttributeWrapperC-expansionPanelDetails-46"
              >
                <div
                  data-testid="line-attribute-x2-input"
                >
                  <div>
                    <label
                      class="MuiFormLabel-root-124"
                    >
                      Thickness
                    </label>
                    <div
                      class="css-10nd86i"
                    >
                      <div
                        class="css-vj8t7z"
                      >
                        <div
                          class="css-1hwfws3"
                        >
                          <div
                            class="css-xp4uvy"
                          >
                            5
                          </div>
                          <div
                            class="css-1g6gooi"
                          >
                            <div
                              class=""
                              style="display: inline-block;"
                            >
                              <input
                                aria-autocomplete="list"
                                autocapitalize="none"
                                autocomplete="off"
                                autocorrect="off"
                                id="react-select-2-input"
                                spellcheck="false"
                                style="box-sizing: content-box; width: 2px; border: 0px; font-size: inherit; opacity: 1; outline: 0; padding: 0px;"
                                tabindex="0"
                                type="text"
                                value=""
                              />
                              <div
                                style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-size: inherit; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                              />
                            </div>
                          </div>
                        </div>
                        <div
                          class="css-1wy0on6"
                        >
                          <span
                            class="css-d8oujb"
                          />
                          <div
                            aria-hidden="true"
                            class="css-1ep9fjw"
                          >
                            <svg
                              aria-hidden="true"
                              class="css-19bqh2r"
                              focusable="false"
                              height="20"
                              viewBox="0 0 20 20"
                              width="20"
                            >
                              <path
                                d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                              />
                            </svg>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
